### ISSUE TITLE
particles are not visible to player after logout and then login

### ISSUE THREAD

**Issue by JurajVajda** [minetest/minetest#4624]
Just came across a weird behavior. After i add particlespawner i can see the particles but after i logout and then login the particles are not visible any more. This is not the case for another players who stayed logged in (didn't logout), those players can still see the particles (unless the logout/login). The server is always running, thus no server restarts are going on. Is this an expected behavior ?

**Comment by nerzhul** [comment]
yes i saw this bug too.
Particle spawners are sent to players at spawn.
Another problem, particlespawner are, if i remember sent to all players, not depending on the distance

**Comment by raymoo** [comment]
I think particle spawners don't send to anyone who joins afterwards. The definition of a particle spawner isn't even stored as far as I can tell, only its lifetime.

**Comment by paramat** [comment]
Particlespawners are usually meant to be fairly short-lived, so no persistence on rejoin is reasonable.

**Comment by JurajVajda** [comment]
Then i dont understand the option to have "infinite lifespan" in particle spawner when you can't see them after you logout/login. In that case is that option obsolete and only way to go is having another ABM running short lived particles every n time.

**Comment by sofar** [comment]
infinite particle spawners are something I haven't seen yet, and they are problematic since if they're really infinite, the player or node they are attached too is guaranteed to be removed at a certain point in time.

I agree this issue is a bug, though.

**Comment by JurajVajda** [comment]
i'm regarding to this option: _time [number] - particle spawner stops after <time> seconds, use 0 for infinite lifespan_

found a work around though, i can save all the positions in the table and reactivate them with LBM as [here](https://github.com/jordan4ibanez/attempted_improvements/blob/master/mods/torches/init.lua).

**Comment by raymoo** [comment]
@sofar Particle spawners don't have to be attached to anything, and they can't be attached to nodes in any sense.

**Comment by sofar** [comment]
No, that will just make multiple particles for players that are on the server longer :)

Best solution is not to use infinite life particle spawners, but regularly respawn them (ideally, node timer based).

@raymoo I didn't mean attached as in "attached through a lua interface", I meant "related to a node or entity" like e.g. smoke particle is related to a torch node.

**Comment by raymoo** [comment]
You could use infinite-life spawners but delete them before respawning them

**Comment by JurajVajda** [comment]
@raymoo yes that's a good idea  :)

**Comment by sofar** [comment]
The risk with doing things like that when players connect is that you're potentially updating hundreds of particle spawners, if you have a lot (e.g. torch smoke)

If you do this through node timers, you're only ever sending a few particle spawners at a time since they are easy to stagger/not happen at the same time.

Of course, torch smoke particles should be implemented using client-side modding, which makes this whole problem go away.

**Comment by JurajVajda** [comment]
@sofar easy solution huh ? :)

yeah so i think it can be reactivated and solved using:
* LBM
* node timers
* combination of both ? 

soon i will apply it for [this mod](https://www.youtube.com/watch?v=5v6VeQyhFnA).

**Comment by raymoo** [comment]
@sofar You can stagger updates from a table too, just you would have to do it manually. I also never said what would be used to trigger the reactivation, you could do it with node timers too, just have a table from node positions to particle spawners that the node uses to refresh. You'd have to be careful to deregister the "spawner refreshing" when the node is removed, though.

For particle spawners attached to players, though, you would be forced to reactivate them on player join, or on a globalstep, or something similar. This might not be too big of an issue, since probably mods wouldn't add lots of spawners to each player, making the number of them limited to the number of players.

Lua entities have the same issues, but you could do it in its on_step callback at some interval.

**Comment by paramat** [comment]
I guess 'infinite lifespan' actually means persist until player leaves, the documentation could perhaps be improved.

**Comment by raymoo** [comment]
Just for clarification, by "reactivate on player join", I mean reactivate whenever any player joins, not just the ones who have particle effects on them.

**Comment by paramat** [comment]
Related #6121

**Comment by paramat** [comment]
1+ year bump, any core dev support?

**Comment by Wuzzy2** [comment]
If this thing is dealt with, please make sure to keep the old behavior by default, i.e. that all particlespawners disappear on shutdown. Because of games that manually respawn particlespawners with e.g. LBMs.
Existing games that use LBM should not be broken by updating this. Thanks for reading.

**Comment by raymoo** [comment]
As far as I can tell, this issue is only for players logging onto a server where a particle spawner was already spawned in that session, not for making them persistent across shutdown.
